File: src/main/java/com/skidders/sigma/managers/ModuleManager.java
package com.skidders.sigma.managers;

import com.google.common.eventbus.Subscribe;
import com.skidders.sigma.events.impl.KeyPressEvent;
import com.skidders.sigma.module.Category;
import com.skidders.sigma.module.Module;
import com.skidders.sigma.module.impl.gui.*;
import net.minecraft.client.MinecraftClient;
import org.lwjgl.glfw.GLFW;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class ModuleManager {

    public List<Module> modules = new ArrayList<>();

    public ModuleManager() {
        modules.add(new ActiveMods());
    }

    @Subscribe
    public void onKey(KeyPressEvent event) {
        if (event.action == GLFW.GLFW_RELEASE && MinecraftClient.getInstance().world != null && MinecraftClient.getInstance().currentScreen == null)
            this.modules.forEach(m -> {
                if (m.key == event.key) {
                    m.setEnabled(!m.enabled);
                }
            });
    }

    public Module getModuleByName(String input) {
        return modules.stream().filter(m -> m.name.equalsIgnoreCase(input)).findFirst().orElse(null);
    }

    public List<Module> getEnabledModules() {
        return modules.stream().filter(m -> m.enabled).collect(Collectors.toList());
    }

    public <V extends Module> V getModuleByClass(Class<V> clazz) {
        Module mod = modules.stream().filter(m -> m.getClass().equals(clazz)).findFirst().orElse(null);

        if (mod == null) {
            return null;
        }

        return clazz.cast(mod);
    }

    public List<Module> getModulesByCategory(Category input) {
        return modules.stream().filter(mod ->
                mod.category.equals(input)).collect(Collectors.toList()
        );
    }

}

File: src/main/kotlin/com/skidders/sigma/managers/ModuleManager.kt
package com.skidders.sigma.managers

import com.google.common.eventbus.Subscribe
import com.skidders.sigma.events.impl.KeyPressEvent
import com.skidders.sigma.module.Category
import com.skidders.sigma.module.Module
import com.skidders.sigma.module.impl.gui.ActiveMods
import net.minecraft.client.MinecraftClient
import org.lwjgl.glfw.GLFW

import java.util.stream.Collectors

class ModuleManager {

    val modules: ArrayList<Module> = arrayListOf()

    constructor() {
        modules.add(ActiveMods())
    }

    @Suppress("UNUSED")
    @Subscribe
    fun onKey(event: KeyPressEvent) {
        if (event.action == GLFW.GLFW_RELEASE && MinecraftClient.getInstance().world != null && MinecraftClient.getInstance().currentScreen == null)
            this.modules.forEach {
                if (it.key == event.key) {
                    it.setEnabled(!it.enabled)
                }
            }
    }

    fun getModuleByName(input: String): Module {
        return modules.stream().filter { it.name.equals(input, ignoreCase = true) }.findFirst().orElse(null)
    }

    fun getEnabledModules(): List<Module> {
        return modules.stream().filter { it.enabled }.collect(Collectors.toList())
    }

    fun <V : Module> getModuleByClass(clazz: Class<V>): V? {
        val mod = modules.stream().filter { it.javaClass.equals(clazz) }.findFirst().orElse(null)

        if (mod == null) {
            return null
        }

        return clazz.cast(mod)
    }

    fun getModulesByCategory(input: Category): List<Module> {
        return modules.stream().filter { it.category == input }.collect(Collectors.toList()
        )
    }

}
